openupgrade_scripts/scripts/mail/16.0.1.10/pre-migration.py

```diff
@@ -97,6 +97,27 @@ def _to_mail_notif_and_email_create_mail_notification_index(env):
         [("name", "=", "to_mail_notif_and_email"), ("state", "=", "to upgrade")]
     )
     if module_to_mail_notif_and_email:
+        openupgrade.logged_query(
+            env.cr,
+            """
+            WITH sub AS(
+                SELECT
+                    id,
+                    mail_message_id,
+                    res_partner_id,
+                    notification_type,
+                    ROW_NUMBER() OVER
+                    (PARTITION BY mail_message_id,res_partner_id,notification_type order by id)
+                    rn
+                FROM mail_notification
+            )
+            DELETE
+            FROM mail_notification mn WHERE EXISTS(
+              SELECT 1 FROM sub
+              where sub.rn > 1 and mn.id = sub.id
+            )
```

Comment:
Cũng ko phải vấn đề gì lắm, nhưng query này sẽ chậm do subquery `SELECT 1 FROM sub where sub.rn > 1 and mn.id = sub.id` phải chạy nhiều lần qua từng row. Có thể dùng query kiểu như sau thì vừa đơn giản vừa tránh đc subquery:
```sql
delete from mail_notification mn1
    using mail_notification mn2
where mn1.id > mn2.id
    and mn1.mail_message_id = mn2.mail_message_id
    and mn1.res_partner_id = mn2.res_partner_id
    and mn1.notification_type = mn2.notification_type;
```

Reply:
@duong77476 thông tin thêm thôi, còn để như cũ cũng ko sao nhé :v

Reply:
Vâng a để e sửa luôn :v